src/main/scala/com/scalableminds/fossildb/FossilDBGrpcImpl.scala

```diff
@@ -33,8 +33,9 @@ class FossilDBGrpcImpl(storeManager: StoreManager) extends FossilDBGrpc.FossilDB
 
   override def put(req: PutRequest) = withExceptionHandler(req) {
     val store = storeManager.getStore(req.collection)
-    require(req.version >= 0, "Version numbers must be non-negative")
-    store.put(req.key, req.version, req.value.toByteArray)
+    val version = req.version.getOrElse(store.get(req.key, None).map(_.version + 1).getOrElse(0L))
```

Comment:
Could `store.get()...` be turned into a function so that it is only executed in the `else` case?

Reply:
Not entirely sure what you mean – `store.get` already is only executed if req.version is None.
A further optimization could be to create a separate `store.getNewestVersionNumber` that doesn’t have to touch the content. Should I do that?